### PR TITLE
interfaces/builtin/firewall_control: allow ufw accesses

### PR DESCRIPTION
**Pull request by mardy**
Fixes: https://bugs.launchpad.net/snapd/+bug/1939711

### CODE REVIEW

**Review by jdstrand — CHANGES_REQUESTED** [review]
I'm going to discuss this from a few angles (cc @alexmurray) to give my thinking on this.

This access feels innocuous but it opens the door for someone to say "I want to drive the system firewalld, shorewall, ..." and then this interface's policy would become overly broad. The purest implementation would create a `ufw-control` interface that allows driving ufw and one for firewalld, shorewall, etc. As a compromise, one could allow this conditionally if `release.OnClassic`, but...

this has been added because microk8s is trying to call ufw, but ufw isn't in the core snaps so they must be shipping it. Since they are shipping it, they could look at the ufw snap to see how it is putting the lock somewhere else. Of course, if you put the lock somewhere else, it can't set the lock on the system so microk8s and the system admin could try to do things at the same time. I question this approach because they might be writing rules out to other locations and two different sets of files and startups would have unpredictable results. They might be trying to use system-files to write out to /etc/ufw like the system one does, but this might cause problems because while ufw is quite stable, there is no guarantee that the ufw they packaged in the snap is compatible with the one on the system (today it is possibly to probably ok to do this with microk8s usage, but imagine if at some future date ufw is updated to write new rules into /etc/ufw/user.rules. The user does this, then installs microk8s, which uses its older ufw that is unable to read the new rules and it errors. Or maybe they use a very new ufw down the line and write out things that break the system ufw). ufw is not designed to have different versions of itself owning and operating on the same data.

This is very similar to why we don't let people use `system-files` or `personal-files` without discussion. If the application is not prepared to handle forward and backward compatibility, write access is inappropriate. I've not reviewed what microk8s is doing on a technical level, but it may want to embed ufw. Instead, I suggest either: a) figure out how to run ufw from hostfs so it is driving the one on the system and then add the ufw lock rule to kubernetes-support, perhaps as a firewall flavor, b) convert its ufw commands to iptables/netfilter commands (setting up microk8s-specific tables so it plays nicely with other software; see ufw, libvirt, lxd, et al (multipass?) for examples) or c) add a snap-fw command to snapd (ala xdg-open) that knows how to drive the host's firewall (ufw, firewalld, etc) and add usage of that command to firewall-control. 'c' is the most ambitious but has a lot of nice properties. A fourth option would be to add a ufw-cli command to ufw that communicates over IPC (eg, DBus) and then create a ufw-control interface that uses it. As upstream ufw, I don't have time to work on that atm, but it is something I've wanted to have for some time (though even if I did, it would be years before that was in widespread use).

Since none of my proposed options includes adding a rule to firewall-control for ufw locking, I'm going to NAK this. I'll defer to @alexmurray on future discussions around this. I do sincerely hope you don't allow applications to muck around in the system's firewall config directly (whether that is ufw, firewalld, shorewall, etc) since that will inevitably cause bugs for you and the upstream maintainers that largely won't be fixable on a timescale your users would want.

**Comment by joedborg** [comment]
Thanks for all the feedback @jdstrand,

Debate of whether or not `firewall-control` should drive the host's `ufw` aside, we're faced with the issue where you can't layout stuff in `/run`.  We're already doing some layout to keep it in the snap (https://github.com/ubuntu/microk8s/blob/feature/jdb/strict/snap/snapcraft.yaml#L434), but the snap won't build if you try to do this with `/run/ufw.lock`.

**Comment by mardy** [comment]
Closing this. I'll be keeping the branch open in case we want to resurrect it in some other form later.

**Pull request closed without merging**
